inc/export/class-endpoint.php

```diff
@@ -0,0 +1,323 @@
+	 */
+	private function get_date_query_arguments( string $date ) {
+		try {
+			$start_date = new DateTime( $date . 'T00:00:00' );
+		} catch ( Exception $e ) {
+			return new WP_Error(
+				'invalid-date',
+				'Could not get query arguments from date argument'
+			);
+		}
+
+		return [
+			'start' => (int) ( $start_date->format( 'U' ) * 1000 ),
+			'end' => ( (int) $start_date->format( 'U' ) + DAY_IN_SECONDS ) * 1000,
+		];
+	}
+
+	/**
+	 * Recursively convert array to CSV string.
+	 *
+	 * @param array $data The data to output as CSV.
+	 * @return string|null
+	 */
+	private function to_csv( array $data ) : ?string {
+		if ( empty( $data ) ) {
+			return null;
+		}
+
+		// Collect column headers.
+		$columns = [];
+
+		// Flatten rows and collect column names.
+		$data = array_map( function ( $row ) use ( &$columns ) {
+			$flattened_data = $this->flatten_array( $row );
+
+			$new_columns = array_keys( $flattened_data );
+			$columns = array_merge( $new_columns, $columns );
+			$columns = array_unique( $columns );
+
+			return $flattened_data;
+		}, $data );
+
+		// Convert to CSV string.
+		$data = array_reduce( $data, function ( $carry, $row ) use ( $columns ) {
+			// Ensure all rows have the same columns.
+			foreach ( $columns as $column ) {
+				$carry .= ( $row[ $column ] ?? '' ) . ',';
+			}
+
+			return trim( $carry, ',' ) . "\n";
+		}, '' );
+
+		// Add header row.
+		$columns = implode( ',', $columns ) . "\n";
+
+		return $columns . $data;
+	}
+
+	/**
+	 * Escape values for CSV.
+	 *
+	 * Ensure commas are preserved and quotes are encoded.
+	 *
+	 * @param mixed $value The value to escape.
+	 * @return mixed
+	 */
+	private function esc_csv( $value ) {
+		if ( is_string( $value ) && strpos( $value, ',' ) !== false ) {
+			if ( strpos( $value, '"' ) !== false ) {
+				$value = str_replace( '"', '""', $value );
+			}
+			$value = sprintf( '"%s"', $value );
+		}
+
+		return $value;
+	}
+
+	/**
+	 * Flattens an array recursively and sets the keys for nested values
+	 * as a dot separated path.
+	 *
+	 * @param array $data The array to flatten.
+	 * @param string $prefix The current key prefix.
+	 * @return array
+	 */
+	private function flatten_array( array $data, string $prefix = '' ) : array {
+		$flattened = [];
+		$prefix = ! empty( $prefix ) ? "{$prefix}." : '';
+
+		foreach ( $data as $key => $value ) {
+			if ( is_array( $value ) ) {
+				$flattened = array_merge( $flattened, $this->flatten_array( $value, "{$prefix}{$key}" ) );
+			} else {
+				$flattened[ "{$prefix}{$key}" ] = $this->esc_csv( $value );
+			}
+		}
+
+		return $flattened;
+	}
+
+	/**
+	 * Lock down the endpoint to only allow administrator users authenticated via application password to access the
+	 * date.
+	 *
+	 * @return bool True if the request has read access for the item, otherwise false.
+	 */
+	public function get_item_permissions_check() {
+		return current_user_can( 'manage_options' );
```

Comment:
Assuming we don't have any meta caps in analytics?

Reply:
Not for analytics directly but we could add an alternative capability here or at least a filter in the meantime.

Reply:
Added a filter for now but we can come back to this. I feel like we'd need a solid solution for role & cap management first.